[test_dispatcher.cpp] | (Runtime) Unit Tests | runtime-unit-tests / runtime_fd2 [bh_p150b_civ2]

## CI Failure Report

<!-- AUTO-TRIAGE-SUMMARY-START -->
**Summary:** [CI] DispatchPagedWriteTestFixture.PagedWrite DRAM tests fail validation in runtime_fd2 [bh_p150b_civ2]
<!-- AUTO-TRIAGE-SUMMARY-END -->

**Workflow:** (Runtime) Unit Tests
**Job:** runtime-unit-tests / runtime_fd2 [bh_p150b_civ2]

### Regression timeline

- **Last passing commit:** `b2e77906b92c` — 2026-07-31T05:14:18Z — https://github.com/tenstorrent/tt-metal/actions/runs/30606247966
- **First failing commit:** `d8ab5b156a26` — 2026-08-04T05:07:35Z — https://github.com/tenstorrent/tt-metal/actions/runs/30879717913/job/91952376495
- **Most recent failing commit:** `056c4c410df4` — 2026-08-05T05:07:09Z — https://github.com/tenstorrent/tt-metal/actions/runs/30977137085/job/92237273127

### Failing job URLs (last 2 runs)

- https://github.com/tenstorrent/tt-metal/actions/runs/30977137085/job/92237273127
- https://github.com/tenstorrent/tt-metal/actions/runs/30879717913/job/91952376495

### Failing test path(s)

- `tests/tt_metal/tt_metal/perf_microbenchmark/dispatch/test_dispatcher.cpp`
- Binary: `./build/test/tt_metal/perf_microbenchmark/dispatch/test_dispatcher --gtest_filter='-*SlowDispatch*'`
- Failing tests:
  - `DispatcherTests/DispatchPagedWriteTestFixture.PagedWrite/page_size16_np512_iter1_DRAM`
  - `DispatcherTests/DispatchPagedWriteTestFixture.PagedWrite/page_size4128_np10_iter1_DRAM`
  - `DispatcherTests/DispatchPagedWriteTestFixture.PagedWrite/page_size16_np13_iter1_DRAM`

### Error excerpt (terminal failure from logs)

**Run 1** (commit `056c4c410df4`, 2026-08-05):
```
../tests/tt_metal/tt_metal/perf_microbenchmark/dispatch/common.h:1376: Failure
Value of: pass
  Actual: false
Expected: true
Dispatcher test failed validation
[  FAILED  ] DispatcherTests/DispatchPagedWriteTestFixture.PagedWrite/page_size16_np512_iter1_DRAM, where GetParam() = 20-byte object <10-00 00-00 00-02 00-00 01-00 00-00 00-00 40-00 01-00 00-00> (546 ms)
[  FAILED  ] DispatcherTests/DispatchPagedWriteTestFixture.PagedWrite/page_size4128_np10_iter1_DRAM, where GetParam() = 20-byte object <20-10 00-00 0A-00 00-00 01-00 00-00 00-00 40-00 01-90 83-47> (563 ms)
[  FAILED  ] DispatcherTests/DispatchPagedWriteTestFixture.PagedWrite/page_size16_np13_iter1_DRAM, where GetParam() = 20-byte object <10-00 00-00 0D-00 00-00 01-00 00-00 00-00 40-00 01-7F 00-00> (564 ms)
[  FAILED  ] 3 tests, listed below:
 3 FAILED TESTS
##[error]Process completed with exit code 1.
```

**Run 2** (commit `d8ab5b156a26`, 2026-08-04):
```
../tests/tt_metal/tt_metal/perf_microbenchmark/dispatch/common.h:1376: Failure
Value of: pass
  Actual: false
Expected: true
Dispatcher test failed validation
[  FAILED  ] DispatcherTests/DispatchPagedWriteTestFixture.PagedWrite/page_size16_np512_iter1_DRAM, where GetParam() = 20-byte object <10-00 00-00 00-02 00-00 01-00 00-00 00-00 40-00 01-00 00-00> (544 ms)
[  FAILED  ] DispatcherTests/DispatchPagedWriteTestFixture.PagedWrite/page_size4128_np10_iter1_DRAM, where GetParam() = 20-byte object <20-10 00-00 0A-00 00-00 01-00 00-00 00-00 40-00 01-00 79-21> (561 ms)
[  FAILED  ] DispatcherTests/DispatchPagedWriteTestFixture.PagedWrite/page_size16_np13_iter1_DRAM, where GetParam() = 20-byte object <10-00 00-00 0D-00 00-00 01-00 00-00 00-00 40-00 01-7F 00-00> (551 ms)
[  FAILED  ] 3 tests, listed below:
 3 FAILED TESTS
##[error]Process completed with exit code 1.
```

### Reproduction steps

Run on a `bh_p150b_civ2` (Blackhole) machine:

```bash
cd <tt-metal-root>
./build/test/tt_metal/perf_microbenchmark/dispatch/test_dispatcher --gtest_filter='-*SlowDispatch*'
```

To run only the failing cases:
```bash
./build/test/tt_metal/perf_microbenchmark/dispatch/test_dispatcher \
  --gtest_filter='DispatcherTests/DispatchPagedWriteTestFixture.PagedWrite/page_size16_np512_iter1_DRAM:DispatcherTests/DispatchPagedWriteTestFixture.PagedWrite/page_size4128_np10_iter1_DRAM:DispatcherTests/DispatchPagedWriteTestFixture.PagedWrite/page_size16_np13_iter1_DRAM'
```

### Notes

- The failures are data-correctness validation failures in the dispatcher's paged write path to DRAM: expected vs. observed 32-bit values mismatch (`common.h:476`, assertion at `common.h:1376`).
- Only DRAM variants fail; L1 variants of the same test pass.
- The same 3 test cases fail across both runs, indicating a regression introduced between `b2e77906b92c` (2026-07-31) and `d8ab5b156a26` (2026-08-04).
- Architecture: Blackhole (`bh_p150b_civ2`).

**Determinism check:**
- Run 1 terminal error: `Dispatcher test failed validation` / `[  FAILED  ] 3 tests` / `##[error]Process completed with exit code 1.` (line 1488–2313 of run1 log)
- Run 2 terminal error: `Dispatcher test failed validation` / `[  FAILED  ] 3 tests` / `##[error]Process completed with exit code 1.` (line 1599–2424 of run2 log)
- Same 3 test names, same failure class, same exit code across both runs.

<!-- AUTO-TRIAGE-METADATA-START -->
`Auto-triage-workflow: (Runtime) Unit Tests`
`Auto-triage-job-name: runtime-unit-tests / runtime_fd2 [bh_p150b_civ2]`
<!-- AUTO-TRIAGE-METADATA-END -->

<!-- AUTO-TRIAGE-INFO-START -->
## About this issue (automated CI auto-triage)
This issue was opened and assigned by an automated AI system. It posts through a bot token, so notifications may look like they come from a person, but they are automated.

The owner is selected from recent commit history and job-ownership data. It reflects who is *likely able to fix or route* this, not necessarily who caused it, so it can be imperfect.

**If this was assigned to the wrong person, you can:**
- Comment `REASSIGN OWNER: <why this is outside your area / which team or test-area it belongs to>` and the system will automatically re-route it to a better-suited owner.
- Unassign yourself (this stops notifications to you).
- Close the issue if it is not a real or relevant failure.

The assigned owner is reminded about every 3 days until the issue is resolved. Full guide: https://tenstorrent.atlassian.net/wiki/spaces/MI6/pages/2545221692/Auto-Triage+Info
<!-- AUTO-TRIAGE-INFO-END -->

--- a/tests/tt_metal/tt_metal/perf_microbenchmark/dispatch/test_dispatcher.cpp
+++ b/tests/tt_metal/tt_metal/perf_microbenchmark/dispatch/test_dispatcher.cpp
@@ -1391,7 +1391,17 @@ INSTANTIATE_TEST_SUITE_P(
         // Testcase: 13 pages x 16 bytes (arbitrary non-even numbers) (L1)
         PagedWriteParams{16, 13, DEFAULT_ITERATIONS_PAGED_WRITE, Common::DRAM_DATA_SIZE_WORDS, false},
         // Testcase: 100 pages x 8192 bytes (high BW) (DRAM)
-        PagedWriteParams{8192, 100, DEFAULT_ITERATIONS_PAGED_WRITE, Common::DRAM_DATA_SIZE_WORDS, true}),
+        PagedWriteParams{8192, 100, DEFAULT_ITERATIONS_PAGED_WRITE, Common::DRAM_DATA_SIZE_WORDS, true},
+        // Testcase: 40 pages x 4112 bytes (DRAM). A page occupies a whole allocation-aligned slot in its bank, so the
+        // in-bank stride between rows of pages is the page size rounded up to that alignment. 4112 is not a multiple
+        // of either DRAM alignment -- 32 B on Wormhole, 64 B on Blackhole -- so the stride differs from the page size
+        // on both, and being larger than a 4 KB dispatch buffer page it also splits across the data the dispatcher has
+        // available, which reaches the row advance on the split-page path as well as the whole-page one. 40 pages
+        // wraps the bank cycle at least three times on either arch.
+        PagedWriteParams{4112, 40, DEFAULT_ITERATIONS_PAGED_WRITE, Common::DRAM_DATA_SIZE_WORDS, true},
+        // Testcase: 40 pages x 4112 bytes (L1). The control for the case above: L1 is 16 B aligned, so the same page
+        // size strides by itself and only the DRAM case can tell the two strides apart.
+        PagedWriteParams{4112, 40, DEFAULT_ITERATIONS_PAGED_WRITE, Common::DRAM_DATA_SIZE_WORDS, false}),
     [](const testing::TestParamInfo<PagedWriteParams>& info) {
         std::stringstream ss;
         ss << "page_size" << info.param.page_size << "_np" << info.param.num_pages << "_iter"
@@ -1557,7 +1567,10 @@ INSTANTIATE_TEST_SUITE_P(
         PagedWriteParams{2048, 128, DEFAULT_ITERATIONS_PAGED_WRITE, Common::DRAM_DATA_SIZE_WORDS, true},
         PagedWriteParams{4128, 10, DEFAULT_ITERATIONS_PAGED_WRITE, Common::DRAM_DATA_SIZE_WORDS, true},
         PagedWriteParams{16, 13, DEFAULT_ITERATIONS_PAGED_WRITE, Common::DRAM_DATA_SIZE_WORDS, true},
-        PagedWriteParams{8192, 100, DEFAULT_ITERATIONS_PAGED_WRITE, Common::DRAM_DATA_SIZE_WORDS, true}),
+        PagedWriteParams{8192, 100, DEFAULT_ITERATIONS_PAGED_WRITE, Common::DRAM_DATA_SIZE_WORDS, true},
+        // Page size that is not a multiple of either DRAM alignment, so the in-bank row stride differs from it; see
+        // the FD instantiation above.
+        PagedWriteParams{4112, 40, DEFAULT_ITERATIONS_PAGED_WRITE, Common::DRAM_DATA_SIZE_WORDS, true}),
     [](const testing::TestParamInfo<PagedWriteParams>& info) {
         return std::to_string(info.param.page_size) + "B_" + std::to_string(info.param.num_pages) + "pages_" +
                std::to_string(info.param.num_iterations) + "iter_" + (info.param.is_dram ? "dram" : "l1");

--- a/tests/tt_metal/tt_metal/perf_microbenchmark/dispatch/test_dispatcher.cpp
+++ b/tests/tt_metal/tt_metal/perf_microbenchmark/dispatch/test_dispatcher.cpp
@@ -70,6 +70,11 @@ struct LinearWriteParams {
     bool is_mcast{};  // Whether to use multicast or unicast
 };
 
+// How the paged write test reads PagedWriteParams::page_size. Random draws a size per run with
+// page_size as the ceiling, so those cases sweep sizes across runs; Exact emits page_size itself,
+// for cases that cover a property of one particular size.
+enum class PageSizeMode : uint8_t { Random, Exact };
+
 // Params that control the data volume, iteration count, and DRAM/L1
 // for the paged write test
 struct PagedWriteParams {
@@ -77,7 +82,8 @@ struct PagedWriteParams {
     uint32_t num_pages{};       // Number of pages
     uint32_t num_iterations{};  // Number of iterations for the test
     uint32_t dram_data_size_words{};
-    bool is_dram{};  // Whether to use DRAM or L1
+    bool is_dram{};                                      // Whether to use DRAM or L1
+    PageSizeMode page_size_mode = PageSizeMode::Random;  // Whether page_size is a ceiling or the size to emit
 };
 
 // Params that control the data volume, iteration count
@@ -350,6 +356,7 @@ class DispatchPagedWriteTestFixture : public BaseDispatchTestFixture,
     uint32_t num_iterations_{};
     uint32_t dram_data_size_words_{};
     bool is_dram_{};
+    PageSizeMode page_size_mode_ = PageSizeMode::Random;
 
     // Get the logical core for this bank
     CoreCoord get_bank_core(uint32_t bank_id) const {
@@ -372,6 +379,7 @@ protected:
         num_iterations_ = p.num_iterations;
         dram_data_size_words_ = p.dram_data_size_words;
         is_dram_ = p.is_dram;
+        page_size_mode_ = p.page_size_mode;
     }
 
 public:
@@ -393,6 +401,16 @@ public:
         // This vector stores commands related information for each iteration
         std::vector<HostMemDeviceCommand> commands_per_iteration;
         const uint32_t page_size_words = page_size_bytes / sizeof(uint32_t);
+
+        // A whole page has to fit in one command for the chunk loop below to make progress. Random
+        // page sizes are bounded by MAX_XFER_SIZE_16B; a PageSizeMode::Exact case is bounded only by
+        // what it asks for.
+        TT_FATAL(
+            page_size_bytes <= max_payload_per_cmd_bytes,
+            "Paged write page size of {} B exceeds the {} B payload one command can carry; lower the page size or "
+            "raise the fetch size",
+            page_size_bytes,
+            max_payload_per_cmd_bytes);
 
         uint32_t remaining_pages = num_pages_;
         uint32_t absolute_start_page = 0;
@@ -460,6 +478,7 @@ public:
     }
 
     uint32_t get_page_size() const { return page_size_; }
+    PageSizeMode get_page_size_mode() const { return page_size_mode_; }
     uint32_t get_num_pages() const { return num_pages_; }
     uint32_t get_num_iterations() const { return num_iterations_; }
     uint32_t get_dram_data_size_words() const { return dram_data_size_words_; }
@@ -486,9 +505,11 @@ public:
         const uint32_t num_banks = device_->allocator_impl()->get_num_banks(buf_type);
         const tt::CoreType core_type = is_dram ? tt::CoreType::DRAM : tt::CoreType::WORKER;
 
-        uint32_t max_allowed = MAX_XFER_SIZE_16B - 1;
-        uint32_t page_size_bytes =
-            payload_generator_->get_random_size(max_allowed, bytes_per_16B_unit, page_size_bytes_param);
+        const uint32_t max_allowed = MAX_XFER_SIZE_16B - 1;
+        const uint32_t page_size_bytes =
+            get_page_size_mode() == PageSizeMode::Exact
+                ? page_size_bytes_param
+                : payload_generator_->get_random_size(max_allowed, bytes_per_16B_unit, page_size_bytes_param);
 
         DeviceCommandCalculator cmd_calc;
         cmd_calc.add_dispatch_write_paged<inline_data_>(page_size_bytes, 0);
@@ -497,7 +518,7 @@ public:
 
         log_info(
             tt::LogTest,
-            "Paged Write test to {} - random page_size: {} bytes, num_pages_per_cmd: {}, iterations: {}",
+            "Paged Write test to {} - page_size: {} bytes, num_pages_per_cmd: {}, iterations: {}",
             is_dram ? "DRAM" : "L1",
             page_size_bytes,
             num_pages_per_cmd,
@@ -1384,8 +1405,10 @@ INSTANTIATE_TEST_SUITE_P(
         PagedWriteParams{2048, 128, DEFAULT_ITERATIONS_PAGED_WRITE, Common::DRAM_DATA_SIZE_WORDS, true},
         // Testcase: 128 pages x 2048 bytes (L1)
         PagedWriteParams{2048, 128, DEFAULT_ITERATIONS_PAGED_WRITE, Common::DRAM_DATA_SIZE_WORDS, false},
-        // Testcase: 10 pages x 4128 bytes (not 4K-aligned) (DRAM)
-        PagedWriteParams{4128, 10, DEFAULT_ITERATIONS_PAGED_WRITE, Common::DRAM_DATA_SIZE_WORDS, true},
+        // Testcase: 10 pages x 4128 bytes (not 4K-aligned) (DRAM). Exact: a smaller drawn size would be 4K-aligned or
+        // no longer straddle a dispatch buffer page, either of which drops the property being covered.
+        PagedWriteParams{
+            4128, 10, DEFAULT_ITERATIONS_PAGED_WRITE, Common::DRAM_DATA_SIZE_WORDS, true, PageSizeMode::Exact},
         // Testcase: 13 pages x 16 bytes (arbitrary non-even numbers) (DRAM)
         PagedWriteParams{16, 13, DEFAULT_ITERATIONS_PAGED_WRITE, Common::DRAM_DATA_SIZE_WORDS, true},
         // Testcase: 13 pages x 16 bytes (arbitrary non-even numbers) (L1)
@@ -1397,11 +1420,15 @@ INSTANTIATE_TEST_SUITE_P(
         // of either DRAM alignment -- 32 B on Wormhole, 64 B on Blackhole -- so the stride differs from the page size
         // on both, and being larger than a 4 KB dispatch buffer page it also splits across the data the dispatcher has
         // available, which reaches the row advance on the split-page path as well as the whole-page one. 40 pages
-        // wraps the bank cycle at least three times on either arch.
-        PagedWriteParams{4112, 40, DEFAULT_ITERATIONS_PAGED_WRITE, Common::DRAM_DATA_SIZE_WORDS, true},
+        // wraps the bank cycle at least three times on either arch. Exact, since every one of those properties is a
+        // property of 4112 specifically.
+        PagedWriteParams{
+            4112, 40, DEFAULT_ITERATIONS_PAGED_WRITE, Common::DRAM_DATA_SIZE_WORDS, true, PageSizeMode::Exact},
         // Testcase: 40 pages x 4112 bytes (L1). The control for the case above: L1 is 16 B aligned, so the same page
-        // size strides by itself and only the DRAM case can tell the two strides apart.
-        PagedWriteParams{4112, 40, DEFAULT_ITERATIONS_PAGED_WRITE, Common::DRAM_DATA_SIZE_WORDS, false}),
+        // size strides by itself and only the DRAM case can tell the two strides apart. Exact for the same reason, and
+        // so the two cases are guaranteed to run the same page size.
+        PagedWriteParams{
+            4112, 40, DEFAULT_ITERATIONS_PAGED_WRITE, Common::DRAM_DATA_SIZE_WORDS, false, PageSizeMode::Exact}),
     [](const testing::TestParamInfo<PagedWriteParams>& info) {
         std::stringstream ss;
         ss << "page_size" << info.param.page_size << "_np" << info.param.num_pages << "_iter"
@@ -1479,7 +1506,8 @@ INSTANTIATE_TEST_SUITE_P(
         // Testcase: 128 pages x 2048 bytes (L1)
         PagedWriteParams{2048, 128, DEFAULT_ITERATIONS_PAGED_WRITE, Common::DRAM_DATA_SIZE_WORDS, false},
         // Testcase: 10 pages x 4128 bytes (not 4K-aligned) (DRAM)
-        PagedWriteParams{4128, 10, DEFAULT_ITERATIONS_PAGED_WRITE, Common::DRAM_DATA_SIZE_WORDS, true},
+        PagedWriteParams{
+            4128, 10, DEFAULT_ITERATIONS_PAGED_WRITE, Common::DRAM_DATA_SIZE_WORDS, true, PageSizeMode::Exact},
         // Testcase: 13 pages x 16 bytes (arbitrary non-even numbers) (DRAM)
         PagedWriteParams{16, 13, DEFAULT_ITERATIONS_PAGED_WRITE, Common::DRAM_DATA_SIZE_WORDS, true},
         // Testcase: 13 pages x 16 bytes (arbitrary non-even numbers) (L1)
@@ -1565,12 +1593,14 @@ INSTANTIATE_TEST_SUITE_P(
     ::testing::Values(
         PagedWriteParams{16, 512, DEFAULT_ITERATIONS_PAGED_WRITE, Common::DRAM_DATA_SIZE_WORDS, true},
         PagedWriteParams{2048, 128, DEFAULT_ITERATIONS_PAGED_WRITE, Common::DRAM_DATA_SIZE_WORDS, true},
-        PagedWriteParams{4128, 10, DEFAULT_ITERATIONS_PAGED_WRITE, Common::DRAM_DATA_SIZE_WORDS, true},
+        PagedWriteParams{
+            4128, 10, DEFAULT_ITERATIONS_PAGED_WRITE, Common::DRAM_DATA_SIZE_WORDS, true, PageSizeMode::Exact},
         PagedWriteParams{16, 13, DEFAULT_ITERATIONS_PAGED_WRITE, Common::DRAM_DATA_SIZE_WORDS, true},
         PagedWriteParams{8192, 100, DEFAULT_ITERATIONS_PAGED_WRITE, Common::DRAM_DATA_SIZE_WORDS, true},
         // Page size that is not a multiple of either DRAM alignment, so the in-bank row stride differs from it; see
         // the FD instantiation above.
-        PagedWriteParams{4112, 40, DEFAULT_ITERATIONS_PAGED_WRITE, Common::DRAM_DATA_SIZE_WORDS, true}),
+        PagedWriteParams{
+            4112, 40, DEFAULT_ITERATIONS_PAGED_WRITE, Common::DRAM_DATA_SIZE_WORDS, true, PageSizeMode::Exact}),
     [](const testing::TestParamInfo<PagedWriteParams>& info) {
         return std::to_string(info.param.page_size) + "B_" + std::to_string(info.param.num_pages) + "pages_" +
                std::to_string(info.param.num_iterations) + "iter_" + (info.param.is_dram ? "dram" : "l1");

--- a/tt_metal/impl/dispatch/kernels/cq_dispatch.cpp
+++ b/tt_metal/impl/dispatch/kernels/cq_dispatch.cpp
@@ -617,10 +617,17 @@ __attribute__((noinline)) void process_write_paged() {
     // Interleaved pages round-robin the banks, so walking them in page order steps the bank index by one and only
     // advances the in-bank offset when it wraps. TensorAccessor::get_noc_addr instead recovers both from the page id
     // every page, at the cost of two magic-multiply divisions by the bank count.
+    //
+    // A page occupies a whole allocator-aligned slot in its bank, so the row stride is the page size rounded up to
+    // that alignment, not the page size itself. The two differ only for a command whose page size is not already
+    // aligned, which no buffer write generates -- EnqueueWriteBuffer sends the aligned page size -- but the command
+    // permits, and test_dispatcher sends.
     constexpr uint32_t num_banks = is_dram ? NUM_DRAM_BANKS : NUM_L1_BANKS;
+    const uint32_t aligned_page_size =
+        align_power_of_2(page_size, interleaved_addr_gen::get_allocator_alignment<is_dram>());
     uint32_t walk_row = interleaved_addr_gen::get_bank_offset_index<is_dram>(page_id);
     uint32_t walk_bank = interleaved_addr_gen::get_bank_index<is_dram>(page_id, walk_row);
-    uint32_t walk_row_addr = base_addr + walk_row * page_size;
+    uint32_t walk_row_addr = base_addr + walk_row * aligned_page_size;
 
     // DPRINT("process_write_paged - pages: {} page_size: {} dispatch_cb_page_size: {}\n", pages, page_size,
     // dispatch_cb_page_size);
@@ -651,7 +658,7 @@ __attribute__((noinline)) void process_write_paged() {
                 page_id++;
                 if (++walk_bank == num_banks) {
                     walk_bank = 0;
-                    walk_row_addr += page_size;
+                    walk_row_addr += aligned_page_size;
                 }
                 data_ptr += page_size;
                 write_length -= page_size;
@@ -682,7 +689,7 @@ __attribute__((noinline)) void process_write_paged() {
             dst_addr_offset = 0;
             if (++walk_bank == num_banks) {
                 walk_bank = 0;
-                walk_row_addr += page_size;
+                walk_row_addr += aligned_page_size;
             }
         }
 

